icu4c/source/test/fuzzer/calendar_fuzzer.cpp

```diff
@@ -0,0 +1,125 @@
+// License & terms of use: http://www.unicode.org/copyright.html
+
+// Fuzzer for ICU Calendar.
+
+#include <cstring>
+
+#include "fuzzer_utils.h"
+
+#include "unicode/calendar.h"
+#include "unicode/localebuilder.h"
+#include "unicode/locid.h"
+
+icu::TimeZone* CreateRandomTimeZone(uint16_t rnd) {
+    icu::Locale und("und");
+    UErrorCode status = U_ZERO_ERROR;
+    std::unique_ptr<icu::StringEnumeration> enumeration(
+        icu::TimeZone::createEnumeration(status));
+    if (U_SUCCESS(status)) {
+        int32_t count = enumeration->count(status);
+        if (U_SUCCESS(status)) {
+            int32_t i = rnd % count;
+            const icu::UnicodeString* id = nullptr;
+            do {
+              id = enumeration->snext(status);
+            } while (U_SUCCESS(status) && --i > 0);
+            if (U_SUCCESS(status)) {
+                return icu::TimeZone::createTimeZone(*id);
+            }
+        }
+    }
+    return icu::TimeZone::getGMT()->clone();
+}
+const char* GetRandomCalendarType(uint16_t rnd) {
+    icu::Locale und("und");
+    UErrorCode status = U_ZERO_ERROR;
+    std::unique_ptr<icu::StringEnumeration> enumeration(
+        icu::Calendar::getKeywordValuesForLocale("calendar", und, false, status));
+    const char* type = "";
+    if (U_SUCCESS(status)) {
+        int32_t count = enumeration->count(status);
+        if (U_SUCCESS(status)) {
+            int32_t i = rnd % count;
+            do {
+              type = enumeration->next(nullptr, status);
+            } while (U_SUCCESS(status) && --i > 0);
+        }
+    }
+    type = uloc_toUnicodeLocaleType("ca", type);
+    return type;
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+    if (size < 2) return 0;
+    uint16_t rnd16;
+    std::memcpy(&rnd16, data, sizeof(rnd16));
+    data += 2;
+    size -= 2;
+    icu::Locale locale = GetRandomLocale(rnd16);
+    const char* type = GetRandomCalendarType(rnd16);
+    std::unique_ptr<icu::TimeZone> timeZone(CreateRandomTimeZone(rnd16));
+    UErrorCode status = U_ZERO_ERROR;
+    icu::LocaleBuilder bld;
+    bld.setLocale(locale);
+    bld.setUnicodeLocaleKeyword("ca", type);
+    locale = bld.build(status);
+    if (U_FAILURE(status)) return 0;
+    std::unique_ptr<icu::Calendar> cal(
+        icu::Calendar::createInstance(*timeZone, locale, status));
+    if (U_FAILURE(status)) return 0;
+    cal->clear();
+
+    while (size > 1 + sizeof(double)) {
+        uint8_t rnd8 = *data++;
+        uint32_t rnd32;
+        double date;
+        std::memcpy(&date, data, sizeof(date));
+        std::memcpy(&rnd32, data, sizeof(rnd32));
+        UCalendarDateFields field = static_cast<UCalendarDateFields>(rnd32 % UCAL_FIELD_COUNT);
+        int32_t srnd32 = (rnd32 % 2 == 0 ? -1 : 1) * rnd32;
+
+        // Shift right 0 - 23 bits to distribute the value more evenly on
+        // different scale.
+        rnd32 >>= (rnd32 % 23);
+        srnd32 >>= (srnd32 % 23);
```

Comment:
You're also tying how far you shift to the original value.

Reply:
I try to make the srnd32 not randomly distributed evenly to the 32bit space, but more evenly in different scale. If I do not do that, the chances of hitting a random very small value is much smaller than hitting a random very large value, for example, there are 10 single digits number can be hit, and 90 double digitis and 900 three ditits number to be hit if we generate randomly without adjustment. Therefore, the code will test 10 times more on the range of 3 digits number than 2 digits number, and 90 times more on  the range of 3 ditis number than a single digit number. I try to shift the bit also based on the random info so the chances of hitting 3 digits random number is closer to hit a 1 digit random number.

Reply:
That's fine, and I agree with you, but a) you're always shifting `rnd32` and `srnd32` the same distance (which maybe isn't really a problem), and b) the random number that determines the actual shift distance is the same random number you're using for other purposes in this test, again meaning two values that should be totally independent of each other have a predictable mathematical relationship.

Reply:
ok, not shifting now.